### PR TITLE
Add `variable.less` to support css variable

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

# 📋 RFC

## Summary

Support CSS Variable to dynamic change antd theme color with static function.

## Basic example
```tsx
import { ConfigProvider } from 'antd';

ConfigProvider.config({
  theme: {
    primaryColor: '#25b864', // Lark green
  },
});
```

## Motivation

In antd, theme is generate on [build time](https://ant.design/docs/react/customize-theme) which means user can not dynamic change theme on runtime. There is some hack way like build replacement css file back of css variable (which may cause wrong design token of same valued variables) or create a less builder on client or server side to build style add inject back.

For the modern browsers, css variable is a cheaper way to dynamic change theme. To avoid breaking current user like IE 11, it should keep all the output as previous version but provide additional build file with css variables version like `antd.variable.css` & `antd.variable.min.css`. So developer can chose to use these new variable style file to enable config provider static function.

## Detailed

To realize this feature, we need create another entry for build. Current components is strong deps on the root style file which use the less function to generate style variables and these function variables can not work in browser. We need:

1. Create a variable entry like `default.less` and replace all the variables by css variable
    * Consider to create script for auto replacement to avoid to variable sync issue
2. Improve ConfigProvider.config to support inject css variables in browser side
3. Move all the components deps apart from the root
    * 🧨 This is still need investigate since components deps on the root which makes recv deps in less. And these makes default version mixed with variable version
4. Refactor related entry apart with target build mode
    * site build entry use variables to support dev requirement
    * dist (webpack) build entry to support multiple mode
    * es/lib build should chose correct build mode

## Drawbacks

Since design token use lots of less function, its hard to replace all the variable with native one. We may lock the static function in alpha for a long time util all the design token is validate to replace

## Adoption strategy

alpha version will release sync with current version of antd. When all the design token replaced, we will release final version.
Not provide less version file since it can not split on less side.

-----

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

Before next major version. We consider to provide css variable support in v4 for the user who do not need support IE.

NOTE: This is far from finish since we do too much on less color calculation. To avoid break current behavior, we need some workaround to make it work.

https://preview-31496-ant-design.surge.sh/components/config-provider-cn/#components-config-provider-demo-theme

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by afc163** [comment]
1. 需要发一段时间的 beta 版本，稳定第一
2. 是否只要自定义 @primary-color 就能更换所有色彩，还是需要 @primary-1 到 10 都要自定义？

**Comment by zombieJ** [comment]
> 1. 需要发一段时间的 beta 版本，稳定第一
> 2. 是否只要自定义 @primary-color 就能更换所有色彩，还是需要 @primary-1 到 10 都要自定义？

1. +1
2. 目前看起来是散装的，大部分的可以合并干掉。几个散装的比较蛋疼。

File: components/button/style/mixin.less [hunk 1]
```diff
@@ -300,12 +329,15 @@
           `
       );
     }
-    & when not (@theme = dark) {
+    & when (not (@theme = dark) and not (@theme = variable)) {
```
Comment (jaredleechn) [inline components/button/style/mixin.less:332]: cool

**Comment by hengkx** [comment]
有没有办法 我只改变`primary color` 就 可以了，自定义颜色太多的话 用户很容易出问题

**Comment by shaodahong** [comment]
> 2. 目前看起来是散装的，大部分的可以合并干掉。几个散装的比较蛋疼。

@zombieJ 散装的 `lighten(@primary-color, 10%)` 这种能支持么

File: components/config-provider/cssVariables.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable import/prefer-default-export, prefer-destructuring */
+
+import { updateCSS } from 'rc-util/lib/Dom/dynamicCSS';
+import { TinyColor } from '@ctrl/tinycolor';
+import { generate } from '@ant-design/colors';
+import { Theme } from './context';
+
+export function registerTheme(theme: Theme) {
```
Comment (chenshuai2144) [inline components/config-provider/cssVariables.tsx:8]: 这个方法能暴露出来，或者挂到 windows 上面吗？

File: components/style/themes/variable.less [hunk 3]
```diff
@@ -0,0 +1,1109 @@
...
+@line-height-base: 1.5715;
+@border-radius-base: 2px;
+@border-radius-sm: @border-radius-base;
+
+// vertical paddings
+@padding-lg: 24px; // containers
+@padding-md: 16px; // small containers and buttons
+@padding-sm: 12px; // Form controls and items
+@padding-xs: 8px; // small items
+@padding-xss: 4px; // more small
```
Comment (afc163) [inline components/style/themes/variable.less:168]:
```suggestion
@padding-xxs: 4px; // more small
```

Comment (afc163) [reply components/style/themes/variable.less:168]: 蛋疼，发现之前就搞错了。

File: components/config-provider/cssVariables.tsx [hunk 4]
```diff
@@ -13,6 +13,7 @@ import {
   excludeParseTokens,
   CalculateValueType,
 } from '../theme';
+import { round } from 'lodash';
```
Comment (afc163) [inline components/config-provider/cssVariables.tsx:16]: oops

**Comment by zombieJ** [comment]
尺寸计算的 design token 和 less 耦合太深拆不出来了。需要从 master 上先拆，免得之后合不起来了。我先把这部分 revert 掉。

**Comment by zombieJ** [comment]
给主题打包做了一个假入口：
* 原本 6 个产物 234,549ms 
* 现在 8 个产物 115,579ms

需要 double confirm 一下产物是否有变化

**Comment by afc163** [comment]
先把 rfc 补一下？

**Comment by zombieJ** [comment]
> 先把 rfc 补一下？

多谢提醒，加上了。

**Comment by afc163** [comment]
Why don't use `<ConfigProvider theme={{ ... }} />`

**Comment by zombieJ** [comment]
> Why don't use `<ConfigProvider theme={{ ... }} />`

ConfigProvider do not create dom node which means we can not create scoped css variable.

**Comment by afc163** [comment]
I afraid `ConfigProvider.config({ theme })` will be deprecated soon in CSS-in-JS solution soon.

**Comment by zombieJ** [comment]
Hmmm. In css-in-js, `<ConfigProvider theme={...} />` will be final way. But for current solution component props may confuse user to think as scoped theme support. Since static function is register variable on the `:root`, it will not break inline scoped theme. We can support this in next major version but not doc in site and recommend developer to replace it.

**Comment by zombieJ** [comment]
@afc163  webpack duplicated version check 有点蛋疼，`rc-cascader` 重构还没完，feature 要再持续挂一段时间

**Comment by M-ZubairAhmed** [comment]
@zombieJ this is an incredible PR, gentle ping on this any update?

**Comment by smshuja** [comment]
This is great. One of the important features for us, is the ability for our clients to customise the theme dynamically. Less.js doesn't work for us and we need the css to be precompiled. Ideally, any less or js functions that are used should be only in default.less file. Any project can then choose to override all the variables and remove references to less/js functions and set pre-computed values. Like `@primary-1`, `@primary-2` etc.

Can't wait for this to get merged.

**Comment by maxim-kolesnikov** [comment]
Great proposal! 🎉 Can we expect these changes in version 5?

**Comment by afc163** [comment]
https://github.com/ant-design/ant-design/issues/30524 一起搞了吧，我不太敢动，怕冲突。

**Review by 07akioni — APPROVED** [review]

**Comment by maxim-kolesnikov** [comment]
Congratulations! 🏆

**Review by xrkffgg — APPROVED** [review]
666

**Review by PeachScript — APPROVED** [review]

File: docs/react/customize-theme-variable.zh-CN.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  </ConfigProvider>
+);
+```
+
+通过静态方法设置主题色以及对应 `prefixCls`：
+
+```ts
+ConfigProvider.config({
+  theme: {
+    prefixCls: 'custom',
```
Comment (afc163) [inline docs/react/customize-theme-variable.zh-CN.md:60]: https://ant.design/components/config-provider-cn/#ConfigProvider.config()-4.13.0+

应该是这样？

```jsx
 ConfigProvider.config({
   prefixCls: 'custom',
 });
```

File: docs/react/customize-theme-variable.zh-CN.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  },
+});
+```
+
+### 编译 less
+
+由于前缀变更，你需要重新生成一份对应的 css 文件。
+
+```bash
+lessc  --modify-var="ant-prefix=custom" antd/dist/antd.variable.less modified.css
```
Comment (afc163) [inline docs/react/customize-theme-variable.zh-CN.md:71]:
```suggestion
lessc --modify-var="ant-prefix=custom" antd/dist/antd.variable.less modified.css
```

File: docs/react/customize-theme-variable.zh-CN.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+## 如何使用
+
+### 引入 antd.variable.min.css
+
+替换当前项目引入样式文件为 CSS Variable 版本：
+
+```diff
+-- import 'antd/dist/antd.min.css';
+++ import 'antd/dist/antd.variable.min.css';
+```
```
Comment (afc163) [inline docs/react/customize-theme-variable.zh-CN.md:21]: 提示一下去掉 babel-plugin-import 里的 css 或 style 属性，避免样式重复（如果用到）。

File: package.json [hunk 8]
```diff
@@ -291,7 +291,7 @@
   "bundlesize": [
     {
       "path": "./dist/antd.min.js",
-      "maxSize": "270 kB"
+      "maxSize": "275 kB"
```
[the remark this answers is not in the document]
Comment (zombieJ) [reply package.json:294]: CSS 计算切换到 js，多了 3+ KB。

File: docs/react/customize-theme-variable.zh-CN.md (hunk 5 above)
Comment (afc163) [reply docs/react/customize-theme-variable.zh-CN.md:60]: 这个还没回复呀。@zombieJ

File: docs/react/customize-theme-variable.zh-CN.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
+---
+order: 7.1
+title: 动态主题（实验性）
+---
+
+除了 [less 定制主题外](/docs/react/customize-theme) 外，我们还提供了 CSS Variable 版本以支持动态切换主题能力。你可以在 [ConfigProvider](/components/config-provider/#components-config-provider-demo-theme) 进行体验。
```
Comment (afc163) [inline docs/react/customize-theme-variable.zh-CN.md:6]: docs/react/customize-theme 里也补一下动态主题内容，链接到这里来。

**Comment by zombieJ** [comment]
纯文档更新， 我在 https://github.com/ant-design/ant-design/pull/31987 顺道一起了

**Comment by maxim-kolesnikov** [comment]
@zombieJ, Thank you for you job! 

Could you explain, how I can use [Dynamic Theme (experience)](https://ant.design/docs/react/customize-theme-variable) in `antd@4.17.0-alpha.0` and `modifyVars` in `webpack.pack.conf` for change size-variables?

That config worked fine..

```ts
// index.ts
import 'antd/dist/antd.variable.less';
```

```js
// webpack.config.js
module.exports = {
  rules: [{
    test: /\.less$/,
    use: [{
      loader: 'style-loader',
    }, {
      loader: 'css-loader',
    }, {
      loader: 'less-loader',
      options: {
        lessOptions: {
          modifyVars: {
            'btn-height-sm': '32px',
            'btn-height-lg': '48px',
            'border-radius-base': '2px',
            // ...other sizes variables
          },
          javascriptEnabled: true,
        },
      },
    }],
  }],
}
```

 .. but, now for  change `primaryColor` in runtime with antd@4.17.0-alpha.0, I should change variable file `.less` to `.css` extension, right? 

```ts
 // index.ts
 -- import 'antd/dist/antd.variable.less';
 ++ import 'antd/dist/antd.variable.min.css';
 
  ConfigProvider.config({
    theme: {
      primaryColor: 'blue',
    },
  });

```

This changes lose modifyVars working. Buttons with size="large" has `height: 40px` as default. 🤔

**Comment by olivewind** [comment]
这个 feature 什么时候可以正式发布？我看现在还是 alpha

**Comment by afc163** [comment]
@olivewind alpha 也可以用起来了。

**Comment by wangjianio** [comment]
升级 4.17.0-alpha.2 报这个错：
```
ERROR in ./node_modules/antd/es/spin/style/css.js 1:0-39
Module not found: Error: Can't resolve '../../style/index-default.css'
```

4.16.13 是好的

**Comment by afc163** [comment]
@wangjianio 提个 issue。

**Comment by arifszn** [comment]
+1. Finally dynamic theming on runtime. It would be nice to see dynamic dark mode implementation in near future too.

**Comment by marklai1998** [comment]
still no dynamic dark mode?

**Comment by dancerphil** [comment]
Great Work! Save us much time on to be compatible with css variables.

**Comment by see2ever** [comment]
modifyVars broken with dynamic theme

**Comment by ckken** [comment]
no dynamic dark mode!

**Pull request merged**